File: LeetCode/1723. Find Minimum Time to Finish All Jobs/solution.py
"""
60 / 60 test cases passed.
Runtime: 64 ms
Memory Usage: 14.5 MB
"""
class Solution:
    def minimumTimeRequired(self, jobs: List[int], k: int) -> int:
        def dfs(workers, idx, limit):
            if idx >= len(jobs):
                return True
            for i in range(len(workers)):
                if workers[i] + jobs[idx] <= limit:
                    workers[i] += jobs[idx]
                    if dfs(workers, idx + 1, limit):
                        return True
                    workers[i] -= jobs[idx]
                if workers[i] == 0 or workers[i] + jobs[idx] == limit:
                    break
            return False

        jobs.sort(reverse=True) 
        l, r = max(jobs), sum(jobs)
        while l < r:
            mid = l + r >> 1
            if dfs([0] * k, 0, mid):
                r = mid
            else:
                l = mid + 1
        return l

File: LeetCode/1723. Find Minimum Time to Finish All Jobs/solution.cpp
/**
 * 60 / 60 test cases passed.
 * Runtime: 0 ms
 * Memory Usage: 50.2 MB 
 */
class Solution {
public: 
    bool dfs(vector<int>& jobs, vector<int>& workers, int i, int limit) {
        if (i >= jobs.size()) {
            return true;
        }
        for (auto& worker: workers) {
            if (worker + jobs[i] <= limit) {
                worker += jobs[i];
                if (dfs(jobs, workers, i + 1, limit)) {
                    return true;
                }
                worker -= jobs[i];
            }
            if (worker == 0 || worker + jobs[i] == limit) {
                break;
            }
        }
        return false;
    }
    int minimumTimeRequired(vector<int>& jobs, int k) {
        sort(jobs.begin(), jobs.end(), greater<int>());
        int l = 0, r = 0;
        for (auto& job: jobs) {
            if (job > l) l = job;
            r += job;
        }
        while (l < r) {
            int mid = (l + r) >> 1;
            vector<int> workerss(k);
            if (dfs(jobs, workerss, 0, mid)) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }
        return l;
    }
};
